spring-vault-core/src/main/java/org/springframework/vault/core/VaultTransitOperations.java

```diff
@@ -123,6 +124,32 @@
 	String encrypt(String keyName, byte[] plaintext, VaultTransitContext transitRequest);
 
 	/**
+	 * Encrypts the provided list of plaintext using the named key.
+	 * The encryption is done using transit backend's batch operation.
+	 *
+	 * works with Vault 0.6.5 and later.
+	 *
+	 * @param keyName must not be empty or {@literal null}.
+	 * @param plaintexts a list of non-empty (or {@literal null}) plaintext.
+	 * @return list of cipher text in the same order as in plaintexts.
+	 */
+	List<Map<String, String>> encrypt(String keyName, List<String> plaintexts);
+
+	/**
+	 * Encrypts the provided list of plaintext using the named key and context.
+	 * The encryption is done using transit backend's batch operation.
+	 *
+	 * works with Vault 0.6.5 and later.
+	 *
+	 * @param keyName must not be empty or {@literal null}.
+	 * @param plaintexts a list of non-empty (or {@literal null}) plaintext.
+	 * @param transitRequests a list of  VaultTransitContext. May be {@literal null} if no request options provided.
+	 * @return list of cipher text in the same order as in plaintexts.
+	 * @throws IllegalArgumentException in case of not matching context found.
+	 */
+	List<Map<String, String>> encrypt(String keyName, List<String> plaintexts, List<VaultTransitContext> transitRequests);
```

Comment:
Same here as above. Additionally, the two lists are not connected – these are just two lists that have no relation to each other. With the single-operation API it wasn't quite obvious but plaintext + context/ciphertext + context have a relation which shouldn't be split up. Maybe we could introduce value objects (`Plaintext`, `Ciphertext`) which encapsulate the actual content along the optional context.

Returning a `List<Ciphertext>` or `List<Encrypted>` (the method before that one was `List<String>`) would express a slightly different context as the caller passes a `List<Plaintext>` to this method. The result could also encapsulate the success/error state of each operation.

Reply:
great suggestion. will take care.

Reply:
How about Value Objects being like:

```java
@Data
class EncryptionRequest {
private String plaintext;
private VaultTransitContext context;
}
```

```java
@Data
class DecryptionRequest {
private String ciphertext;
private VaultTransitContext context;
}
```

A version of below APIs could be created to leverage above VOs:

```java
public String encrypt(String keyName, byte[] plaintext,
			VaultTransitContext transitRequest);
//newer version
public String encrypt(EncryptionRequest encryptionRequest);

```

```java
	public byte[] decrypt(String keyName, String ciphertext,
			VaultTransitContext transitRequest);

//newer version
public byte[] decrypt(DecryptionRequest decryptionRequest);

```

Reply:
`EncryptionRequest` should carry a binary payload. `byte[]` or `ByteBuffer` should be fine. The types should be immutable and expose factory methods like `of(String)` and `of(byte[])`. The context can be associated via a `with(VaultTransitContext)` creating a new object. I think a builder would not be a good fit here.

I'm also a bit concerned regarding the name. Usually, all `…Request` objects encapsulate the whole request. For batching, we don't want to include the key in the request because that would be confusing which key to use.

Please don't change existing API because it would break backward compatibility.

Reply:
great suggestions on the factory pattern.
yes I wasn't suggesting to change existing APIs but was suggesting to create parallel APIs with new signature but eventually doing the same. I wouldn't do it as its not in batch api scope.